### ISSUE TITLE
Add MaxWaitingTime in case of job starvation of lower priority

### ISSUE THREAD

**Issue by kerthcet** [kubernetes-sigs/kueue#754]
We enqueue the jobs via the priorities mainly, which will cause job starvation of lower priorities，we can provide a mechanism to avoid this.

**What would you like to be added**:

A new field MaxWaitingTime to avoid job starvation. 

**Why is this needed**:

**Completion requirements**:

This enhancement requires the following artifacts:

- [x] Design doc
- [x] API change
- [x] Docs update

The artifacts should be linked in subsequent comments.

**Comment by alculquicondor** [comment]
what would happen when the workload hits the maxWaitingTime?

**Comment by kerthcet** [comment]
It will be popped out for scheduling whatever the priority.

**Comment by tenzen-y** [comment]
> It seems a bit aggressive. Another way this can be modeled is that the priority actually goes up.

I like this.

> Another important question is at which object this is controlled. Workload? ClusterQueue? global?

Maybe, ClusterQueue?
If Workload control this, it means batch users can modify this.

**Comment by kerthcet** [comment]
> Only moving the workload to the head of the queue will be pointless when preemption is enabled, it will just be evicted at the next scheduler cycle.

We're not talking about preemption, just in case low priority jobs are starved by continuous high priority jobs.

**Comment by Gekko0114** [comment]
Hi,
It looks interesting, can I work on this issue?
If it is ok, I will clarify how to implement it later.

**Comment by kerthcet** [comment]
Yes, you can take it if you like, but we haven't reached an agreement about the implementation, like a maxWaitingTime, or increate the priority per scheduling cycle, or something else. And we should also consider about the preemption.

**Comment by Gekko0114** [comment]
Sure, I will read this thread in detail and clarify it.
If I get an agreement, I will assign it to me.

**Comment by Gekko0114** [comment]
Hi @alculquicondor,

Thanks for your response.

I have a question.
Why do you think it's better to add a separate object `WorkloadPriorityClass` rather than adding a new field to `clusterQueue`? Is it because it would be easier for batch users to manage priority by choosing a PriorityClass for each job?

> what could another action be?

I think we should consider implementing another action going to the head of the ClusterQueue if it times out, which was the initial idea @kerthcet mentioned.

> how to keep record of the last time the priority was increased? Do we actually update the priority in the object or we just track in memory?

I don't have a strong opinion on this, but I think it's better to update the priority in the object. If so, admins can easily change the priority manually.

> Definitely worth a KEP:

I agree. I will create it.

**Comment by alculquicondor** [comment]
> Why do you think it's better to add a separate object WorkloadPriorityClass rather than adding a new field to clusterQueue? Is it because it would be easier for batch users to manage priority by choosing a PriorityClass for each job?

Yes, I don't think it's realistic that all jobs in a ClusterQueue would have the same behavior.

> I think we should consider implementing another action going to the head of the ClusterQueue if it times out, which was the initial idea @kerthcet mentioned.

I'm not too sure. Because this workload could be immediately preempted in the next iteration. It's better if it's priority changes.

> I don't have a strong opinion on this, but I think it's better to update the priority in the object. If so, admins can easily change the priority manually.

Yes, but also to my point above about not being easily preempted.
Another thing to think about is how do we know when the priority was bumped for the last time. In memory is not enough, because the controller could be restarted.

Can you look around in other queueing systems  if they have dynamic priority and how it works?

**Comment by Gekko0114** [comment]
> I'm not too sure. Because this workload could be immediately preempted in the next iteration. It's better if it's priority changes.
> Yes, but also to my point above about not being easily preempted.

I have one question about preemption.
If we update the priority in the object and a pod starts running, my understanding is that the pod would not be easily preempted since it would have a higher priority. Is that correct?

> Another thing to think about is how do we know when the priority was bumped for the last time. In memory is not enough, because the controller could be restarted.

I agree. It's important.

> Can you look around in other queueing systems if they have dynamic priority and how it works?

Sure. I will look around and rethink the design based on your explanation.

**Comment by alculquicondor** [comment]
> If we update the priority in the object and a pod starts running, my understanding is that the pod would not be easily preempted since it would have a higher priority. Is that correct?

Exactly. That's probably desired? Which means we should probably add a `maxPriority` field in the priorityClass.
So we have `base`, `step`, `delayForStep`, and `max` fields (exact names TBD).

**Comment by Gekko0114** [comment]
Sounds good!

I wasn't quite clear on the intention behind `step` and `maxPriority`.

So, if I understand correctly, `base` represents the priority when the job is initially created, `delayForStep` indicates the time (e.g., 1 hour) to wait before changing the priority after a timeout.
`step` determines the adding priority to set when a timeout occurs, and `maxPriority` is the priority used to override and prevent preemption of actively running pods, is that correct?

**Comment by alculquicondor** [comment]
Almost there. So after 1h, we have `p = base + step`. What happens after 2h? My thinking is that `p = base + step * 2`. Then we need a `maxPriority`, as we don't want it to go to infinity. By default we can probably cap it at MAX_INT32.

**Comment by kerthcet** [comment]
I think we can leverage the workload priority to mitigate this issue. The general idea would like when waiting time ends, we'll raise the workload priority to a higher value and job will enqueue. Next scheduling cycle, the job will not be preempted as designed.

cc @B1F030

**Comment by kerthcet** [comment]
For this case, maybe we can be more straightforward here like: 
1. Set the maxWaitingTime in localQueue/clusterQueue/Job
2. When hit the deadline, raise the priority to the top of the queue

Intuitive and easy to control the job queueing.

**Comment by B1F030** [comment]
Here I drawed a simple design of this mechanism:

![image](https://github.com/kubernetes-sigs/kueue/assets/77265354/785fcc09-22d3-4c3f-88c9-f9a1d06d758f)

The `prototype` represents that there is a `basePriority`, when this workload hits the `maxWaitingTime`, we just simply change its priority to `MAX_INT32`.

And based on the `prototype`, here I have another design:
First we use the workload's priority as `basePriority`, and after `delayTime`(for example, 30m), add its priority by `step`(100), then after another `delayForStep`(10m), add step again.
Until the workload hits the `maxWaitingTime`, its priority will be add up straight to the `maxPriority`(MAX_INT32).

But the second design may be a little complex, and I think we should not expose the `step` and `delayForStep` to user in alpha version. We can define it properly, provide a default policy, that will be easy to use.

What do you think?

**Comment by alculquicondor** [comment]
Let's take a step back... do you have a real requirement from an end-user that you can share? What is the behavior they expect, regardless of how the API looks like.

To me, the idea of a Job getting the highest priority after some timeout sounds odd. Is the job really the most important of all? Additionally, this implies that the job is potentially never going to be preempted. Is this acceptable? Have you looked into older systems for inspiration?

**Comment by kerthcet** [comment]
> But the second design may be a little complex, and I think we should not expose the `step` and `delayForStep` to user in alpha version. We can define it properly, provide a default policy, that will be easy to use.
> 
> What do you think?

Complex + 1 .., let's discuss the rationality first.
